dataplane/grpc/grpc_dataplane.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 Cisco Systems Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/projectcalico/cni-plugin/dataplane/grpc/proto"
+	"github.com/projectcalico/cni-plugin/pkg/types"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	timeout = 5 * time.Second
+)
+
+type grpcDataplane struct {
+	socket            string
+	allowIPForwarding bool
+	mtu               int
+	logger            *logrus.Entry
+}
+
+func NewGrpcDataplane(conf types.NetConf, logger *logrus.Entry) (*grpcDataplane, error) {
+	socket, ok := conf.DataplaneOptions["socket"].(string)
+	if !ok {
+		return nil, fmt.Errorf("GRPC dataplane socket not configured")
+	}
+	return &grpcDataplane{
+		socket:            socket,
+		allowIPForwarding: conf.ContainerSettings.AllowIPForwarding,
+		mtu:               conf.MTU,
+		logger:            logger,
+	}, nil
+}
+
+func (d *grpcDataplane) DoNetworking(
+	args *skel.CmdArgs,
+	result *current.Result,
+	desiredVethName string,
+	routes []*net.IPNet,
+) (ifName, contTapMAC string, err error) {
+	d.logger.Infof("Connecting to GRPC backend server at %s", d.socket)
+	conn, err := grpc.Dial(d.socket, grpc.WithInsecure())
+	if err != nil {
+		return "", "", fmt.Errorf("cannot connect to grpc dataplane: %v", err)
+	}
+
+	c := proto.NewCniDataplaneClient(conn)
+
+	request := &proto.AddRequest{
+		InterfaceName:            args.IfName,
+		Netns:                    args.Netns,
+		DesiredHostInterfaceName: desiredVethName,
```

Comment:
The desired desiredVethName could be, and will be if the orchestrator is not k8s, empty. The "linuxDataplane" defaults this to
`hostVethName = "cali" + args.ContainerID[:utils.Min(11, len(args.ContainerID))]`. I suggest factoring that out and using it here as a default.

Reply:
I moved this default name computation to plugin.go to use the same default value for non-k8s orchestrators for all dataplanes. This ensures DoNetworking never receives an empty value in desiredVethName.

I also modified the k8s code to take into account the name returned by DoNetworking, as not all dataplanes may be able to comply with the requested name (this is the case for VPP in particular, where we don't get to choose the name of the interface on the VPP side).

Reply:
Good thinking.